File: Linked List/Doubly Linked List/Doubly Circular/DC.java
// Doubly circular linked list

import java.Util.Scanner;

class node
{
    public int data;
    public node next;
    public node previous;

    public node(int no)
    {
        data=no;
        next=null;
        previous=null;
    }
}

class Dcircular
{
    public node head ;
    public node tail;

    public Dcircular()
    {
        head =null;
        tail=null;
    }

    public void InserFirst(int no)
    {
        node newn= new node(no);

        if(head == null && tail == null)
        {
            head = newn;
            tail= newn;
        }
        else
        {
            newn.next = head;
            head.next.previous=newn;
            head = newn;
        }
        head.previous=tail;
        tail.next =head;
    }

    public void InsertLast(int no)
    {
        node newn=new node(no);

        if(head == null && tail == null)
        {
            head = newn;
            tail= newn;
        }
        else
        {
            tail.next = newn;
            tail.next.previous=tail;
            tail=newn;
        }
        tail.next =head;
        head.prevoius=tail;
    }

    public void DeleteFirst()
    {
        if(head == null && tail == null)
        {
            return 0;
        }
        else if(head == tail)
        {
            head = null;
            tail = null;
            System.gc();
        }
        else
        {
            head = head.next;
            System.gc();
            head.previous=tail;
            tail.next =head;
        }
    }

    public void DeleteLast()
    {
        if(head == null && tail == null)
        {
            return 0;
        }
        else if(head == tail)
        {
            head = null;
            tail =null;
            System.gc();
        }
        else
        {
            node temp=head;

            while(temp.next.next != tail)
            {
                temp=temp.next;
            }
            temp.next = tail;
            System.gc();
            tail.next =head;
            head.previous=tail;
        }
    }

    public void Display()
    {
        node temp=head;

        while(temp != tail)
        {
            System.out.println("|"+temp.data+"|<=>");
            temp=temp.next;
        }
    }
}

class DC 
{
    public static void main(String a[])
    {
        Scanner sc = new Scanner(System.in);

        Dcircular dc = new Dcircular();

        System.out.println("Insert elements at first:");
        dc. InsertFirst(30);
        dc.InsertFirst(20);
        dc.InsertFirst(10);
        dc.Display();
        System.out.println("\n-----------------------------------------");

        System.out.println("Insert elements at Last:");
        dc.InsertLast(300);
        dc.InsertLast(400);
        dc.Display();
        System.out.println("\n-----------------------------------------");

        System.out.println("Delete Element at First:");
        dc.DeleteFirst();
        dc.Display();
        System.out.println("\n-----------------------------------------");

        System.out.println("Delete Element at Last:");
        dc.DeleteLast();
        dc.Display();
        System.out.println("\n-----------------------------------------");
    }
}

File: Linked List/Doubly Linked List/Doubly Circular/DC.c
// Doubly Circular Linked List

#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node *next;
    struct node *previous;
};

typedef struct node NODE;
typedef struct node *PNODE;
typedef struct node **PPNODE;

int InsertFirst(PPNODE head,PPNODE tail,int no)
{
    PNODE newn=NULL;
    newn=(PNODE)malloc(sizeof(NODE));
    newn->data=no;
    newn->next =NULL;
    newn->previous =NULL;

    if((*head)== NULL && (*tail) ==NULL)
    {
        *head = newn;
        *tail=newn;
    }
    else
    {
        newn->next = *head;
        *head = newn;
    }
    (*tail)->next =*head;
    (*head)->previous = *tail;
}

int InsertLast(PPNODE head,PPNODE tail,int no)
{
    PNODE newn=NULL;
    newn=(PNODE)malloc(sizeof(NODE));
    newn->data=no;
    newn->next =NULL;
    newn->previous =NULL;

    if((*head)== NULL && (*tail) ==NULL)
    {
        *head = newn;
        *tail=newn;
    }
    else
    {
        (*tail)->next = newn;
        *tail = newn;
    }
    (*tail)->next =*head;
    (*head)->previous = *tail;
}

int DeleteFirst(PPNODE head,PPNODE tail)
{
    if((*head)== NULL && (*tail) ==NULL)
    {
       return 0;
    }
    else if(*head == *tail)
    {
        free(*head);
    }
    else
    {
        PNODE temp=*head;

        *head = (*head)->next;
        free(temp);
        (*tail)->next =*head;
        (*head)->previous = *tail;
    } 
}

int DeleteLast(PPNODE head,PPNODE tail)
{
    if((*head)== NULL && (*tail) ==NULL)
    {
       return 0;
    }
    else if(*head == *tail)
    {
        free(*head);
    }
    else
    {
        PNODE temp=*head;

        while(temp->next != (*tail))
        {
            temp=temp->next;
        }
        free(temp->next);
        *tail = temp;
        (*tail)->next =*head;
        (*head)->previous = *tail;
    } 
}

int InsertAtPosition(PPNODE head,PPNODE tail,int no,int ipos)
{
    if(ipos >=1 || ipos < iCountNode+1)
    {
        return 0;
    }

    if(ipos ==1)
    {
       InsertFirst(head,tail,no);
    }
    else if(ipos == iCountNode+1)
    {
        InsertLast(head,tail,no);
    }
    else
    {
        PNODE temp=*head;

        for(int i=-;i<ipos-1;i++)
        {
            temp = temp->next;
        }
        newn->next = temp->next;
        temp->next->previous= newn;
        temp->next = newn;
        newn->previous = temp;
    } 
}

int DeleteAtPosition(PPNODE head,PPNODE tail,int ipos)
{
    if(ipos >=1 || ipos < iCountNode)
    {
        return 0;
    }

    if(ipos ==1)
    {
       DeleteFirst(head,tail);
    }
    else if(ipos == iCountNode)
    {
        DeleteLast(head,tail);
    }
    else
    {
        PNODE temp=*head;
        PNODE target = NULL;

        for(int i=-;i<ipos-1;i++)
        {
            temp = temp->next;
        }
        target = temp->next;
        temp->next = target->next;
        free(target);
        temp->next->previos =temp;
    } 
}

int Display(PNODE head,PNODE tail)
{
    PNODE temp=head;
    do
    {
        printf("|%d|<=>",temp->data);
        temp=temp->next;
    }while(temp != tail->next);
}

int main()
{
    PNODE first = NULL;
    PNODE last = NULL;
    int ipos=0,idel=0;

    InsertFirst(&first,&last,30);
    InsertFirst(&first,&last,20);
    InsertFirst(&first,&last,10);
    printf("Insert Elements at First:");
    Display(first,last);

    printf("\n\n---------------------------------------------------");
    InsertLast(&first,&last,90);
    printf("\n\nInsert Elements at Last:");
    Display(first,last);

    printf("\n\n---------------------------------------------------");
    DeleteFirst(&first,&last);
    printf("\n\nDelete First Elements:");
    Display(first,last);

    printf("\n\n---------------------------------------------------");
    DeleteLast(&first,&last);
    printf("\n\nDelete Last Elements:");
    Display(first,last);

    printf("\n\n---------------------------------------------------");
    InsertAtPosition(&first,&last,ipos,800);
    printf("\n\nInsert Elemets at Position:");
    printf("\nEnter Position:");
    scanf("%d",&ipos);
    Display(first,last);

    printf("\n\n---------------------------------------------------");
    DeleteAtPosition(&first,&last,ipos);
    printf("\n\nDelete Elemets at Position:");
    printf("\nEnter Position:");
    scanf("%d",&idel);
    Display(first,last);

    return 0;
}